Jinja report

simoess opened the pull request:
Add a new html report  #164 

This is a draft PR to include a new jinja2 report to Arche.

The Report class was been refactored to permit multiples report types/styles.

Remaining tasks
- [x] Remove icon when there's more data to show
- [x] Clickable links opens in a new tab
- [x] Fix plots
- [x] Fix data_quality_report
- [x] Fix tests
- [x] Add styles for error messages
- [x] Add styles for dataframes
- [x] Create a base template with all the macros
- [x] Create two separate templates
  - [x] full report
  - [x] individual report

simoess commented:
@manycoding Please let me know what do you think about this solution. Should I proceed to add the plots?

manycoding commented:
@simoess  I looked through, but before I start actual reviewing:
1. Why you kept the old report? I think there's no need to have both. I'd prefer to have markdown if it's possible to render with jinka2, otherwise let's keep HTML.
I'd be nice to see how it actually looks here. Or this pr is about the code, not the style?

simoess commented:
@manycoding  I updated the code and removed the Markdown report.

This is a screenshot of the report. The report style can be easily changed by altering the template.
![arche_jinja_report_print](https://user-images.githubusercontent.com/19392851/65333253-85549100-db96-11e9-955d-c569b86eb9c9.png)

manycoding commented on src/arche/data_quality_report.py:
```diff
@@ -7,7 +7,7 @@
 from arche.readers.items import CloudItems
 from arche.readers.schema import Schema
-from arche.report import Report
+from arche.reports import Report
```
I guess it doesn't work

manycoding commented on src/arche/report.py:
```diff
@@ -1,102 +1,66 @@
-            cls.write_color_text(msg, color=Fore.YELLOW)
-        elif outcome == Outcome.PASSED.name:
-            cls.write_color_text(msg, color=Fore.GREEN)
+    def _rule_outcome(self, rule) -> str:
```
typing

manycoding commented on src/arche/report.py:
```diff
@@ -1,102 +1,66 @@
+                       )
+        return [rule[1] for rule in rules]
+
+    def display(self) -> None:
```
`__call__()`

manycoding commented on src/arche/report.py:
```diff
@@ -1,102 +1,66 @@
+            rule_outcome=self._rule_outcome,
+            pd=pd,
+        )
+        display(HTML(resultHTML), metadata={"isolated": True})
```
I think there's `display_html`

manycoding commented on src/arche/report.py:
```diff
@@ -1,102 +1,66 @@
+            f.figures
+
+        template = self.env.get_template('template.html')
+        resultHTML = template.render(
```
`rendered` seems better for me

// TODO async output https://jinja.palletsprojects.com/en/2.10.x/api/#jinja2.Template.render_async. I guess we need to make all rules async first. (just a thought, not necessary for this pr)

simoess commented on src/arche/arche.py:
```diff
@@ -123,12 +123,10 @@ def get_items(
-        self.report.write_summaries()
-        self.report.write("\n" * 2)
-        self.report.write_details(short)
+        self.report.display()
```
I changed the report class to be a callable.

manycoding commented on src/arche/rules/result.py:
```diff
@@ -80,6 +92,21 @@ def info(self):
+        elif any([level.value == Level.WARNING.value for level in message_level]):
+            return Outcome.WARNING
+        else:
+            return Outcome.SKIPPED
```
Docstring is obvious, let's just remove it.

The rest of the code could be simpler
```
if not self.messages:
    return Outcome.PASSED
 if Level.ERROR in self.messages.keys():
    return Outcome.FAILED
if Level.WARNING in self.messages.keys():
    return Outcome.WARNING
return Outcome.SKIPPED
```

manycoding commented on src/arche/rules/result.py:
```diff
@@ -158,14 +185,13 @@ def get_detailed_messages_count(self):
-        for f in self.figures:
-            f.show()
+        report.save(self)
+        report(self)
```
```
IPython.display.clear_output()
Report(self)
```

simoess replied:
> ```
> IPython.display.clear_output()
> Report(self)
> ```

It's not possible to call Report(self) (this will call the constructor). I changed it to Report()(self)

manycoding commented on tests/test_arche.py:
```diff
@@ -176,9 +173,8 @@ def test_report_all(mocker, get_cloud_items):
-    mocked_write_summaries.assert_called_once_with(a.report)
-    mocked_write.assert_called_once_with("\n" * 2)
     assert executed == a.report.results.keys()
+    mocked_call.assert_called_once()
```
```suggestion
    mocked_call.assert_called_once_with(a.report)
```

manycoding commented on tests/test_arche.py:
```diff
@@ -225,19 +221,19 @@ def test_run_all_rules_collection(mocker, get_collection_items):
     a.validate_with_json_schema()
 
-    mocked_show.assert_called_once_with(res)
+    mocked_call.assert_called_once()
```
```suggestion
    mocked_call.assert_called_once_with(res)
```

manycoding commented on tests/test_report.py:
```diff
@@ -30,103 +31,49 @@
-def test_write_details(mocker, get_df, capsys, messages, expected_details):
-    mock_pio_show = mocker.patch("plotly.io.show", autospec=True)
-    md_mock = mocker.patch("arche.report.display_markdown", autospec=True)
+def test_display(mocker, get_df, capsys, messages, expected_details):
```
```suggestion
def test_report_call(mocker, get_df, capsys, messages, expected_details):
```

manycoding commented on tests/test_report.py:
```diff
@@ -30,103 +31,49 @@
-    Report.write_rule_details(outcome)
-    assert not capsys.readouterr().out
+    r = Report()
+    r(outcome)
```
```suggestion
   Report(outcome)
```

manycoding commented on tests/test_report.py:
```diff
@@ -30,103 +31,49 @@
     [({Level.INFO: [("summary", "very detailed message")]}, "very detailed message")],
 )
-def test_write_rule_details(capsys, message, expected_details):
+def test_write_rule_details(mocker, get_df, capsys, message, expected_details):
```
```suggestion
def test_report_call_argument(mocker, get_df, capsys, message, expected_details):
```
Would be nice if we remove `@pytest.mark.parametrize(` since we have only one set.

manycoding commented on tests/test_report.py:
```diff
@@ -30,103 +31,49 @@
-    md_mock = mocker.patch("arche.report.display_markdown", autospec=True)
-    Report.write_detailed_errors(errors, short, keys_limit)
-    md_mock.assert_has_calls(mocker.call(m) for m in expected_messages)
+def test_display_errors(mocker, get_job_items, get_schema):
```
```suggestion
def test_report_call_with_errors(mocker, get_job_items, get_schema):
```

simoess commented on src/arche/rules/result.py:
```diff
@@ -25,6 +25,18 @@ class Outcome(Enum):
     FAILED = 2
+    WARNING = 3
+
+    def __eq__(self, other):
```
To compare a Outcome with another Outcome. The comparison doesn't work without it

simoess replied:
> We are adding comparison operators for what?

To compare a Outcome with another Outcome. The comparison doesn't work without it

simoess commented on src/arche/arche.py:
```diff
@@ -124,12 +124,10 @@ def get_items(
     def save_result(self, rule_result):
         self.report.save(rule_result)
 
-    def report_all(self, short: bool = False) -> None:
```
I added a parameter on the macros.

simoess commented:
> 1. Graphs are not displayed. Again.
>    https://arche.readthedocs.io/en/jinja_report/nbs/Rules.html

Solved

simoess commented on tests/rules/test_result.py:
```diff
@@ -72,14 +71,21 @@ def test_result_err_keys(messages, true_err_keys):
+
+
+def test_outcome_equality():
+    outcomes = [Outcome.SKIPPED, Outcome.PASSED, Outcome.FAILED, Outcome.WARNING]
```
> Good, but the simplest way:
> 
> ```
> assert Outcome.SKIPPED == Outcome.SKIPPED
> assert Outcome.WARNING != Outcome.PASSED
> ```
> 
> I love the simplest solution which does the job because it's easier to read. It's up to you if you want to change the tests or not.

Yes, but doing so wouldn't the code need to have several lines in order to test every possible pair of values?

manycoding replied:
We don't have to test every possible pair. The goal is to test `__eq__ and __ne__` operators, so what can go wrong:
1. anyone of them is not implemented
2. anyone of them is implemented wrong
Thus, if we have at least one use case to test both operators, they work. I.e. if there's `Outcome.SKIPPED == Outcome.SKIPPED` we know all others work (since they all belong to the single class`, same with `!=`.
Makes any sense?

When I write a unit test, I am thinking what can go wrong from human side (where the developer might make an error). It's either the function doesn't return what's expected or doesn't cover all use cases.

simoess commented:
> 1. I see only big blank space after each rule

Now the height of the output is auto adjusted

manycoding commented:
@simoess The last thing left - add a line to changes https://github.com/scrapinghub/arche/blob/jinja_report/CHANGES.md#037dev-work-in-progress. From the user perspective
@Gallaecio @ejulio @andersonberg Hi, could you please review?

ejulio commented on src/arche/report.py:
```diff
@@ -1,105 +1,51 @@
-        elif outcome == Outcome.PASSED.name:
-            cls.write_color_text(msg, color="#0B6121")
+    def __call__(self, rule: Result = None, keys_limit: int = None) -> None:
+        if not rule:
```
This could be `if rule`

ejulio commented on src/arche/rules/result.py:
```diff
@@ -80,6 +94,25 @@ def info(self):
 
+    @property
+    def outcome(self) -> Outcome:
+        if not self._outcome:
```
These conditions could go in a separate method, so it would be clear they are getting an `Outcome` and setting `_outcome` if it is not set yet.

ejulio replied:
What happens if I set an `outcome`, say `PASSED`, but added an error later?

manycoding replied:
My idea is to set outcome based on messages' levels and allow to override it.
If `_outcome` has the value, it'll returned notwithstanding what else is in messages.

simoess replied:
@ejulio I agree with @manycoding. There's the default behavior of defining the outcome based on the message log levels; In the case that the rule has some exception and the outcome can't be defined based on the messages alone, the rule can override it.

ejulio commented on src/arche/rules/result.py:
```diff
@@ -80,6 +94,25 @@ def info(self):
+        return self._outcome
+
+    @outcome.setter
+    def outcome(self, outcome):
```
Probably this is an internal behavior and shouldn't be set outside the class

manycoding replied:
Sometimes we want to override outcome, that's why there's a property.

simoess replied:
> @simoess The last thing left - add a line to changes https://github.com/scrapinghub/arche/blob/jinja_report/CHANGES.md#037dev-work-in-progress. From the user perspective

Done

simoess replied:
@ejulio I agree with @manycoding  on this one.

ejulio commented on src/arche/report.py:
```diff
@@ -1,105 +1,51 @@
-            display_markdown(f"{len(keys)} items affected - {attribute}: {sample}")
+            template = self.env.get_template("single-rule.html")
+            resultHTML = template.render(
+                rule=rule,
```
Couldn't it be just a rule wrapped in a list or something like that?
So, there is only one template (multiple entries) and this is a special case (where you wrap a value in a list.

manycoding replied:
But the information is different, for single rule we add `summary` which goes into a separate report for all rules.

simoess replied:
@ejulio I don't think so. The information is displayed in a different way by using the single rule(the plots are displayed directly on the rule and not on a separated "plot" tab as on the full report.

ejulio commented:
Just not sure if there should be some update in the docs.
Images/Examples?
From my understanding, the reporting API is "private", right?

manycoding commented:
@ejulio There's a docs branch built https://arche.readthedocs.io/en/jinja_report/ with all the changes. `show()` and `report_all()` are documented, and that's all API.

@simoess The graph in the main report is blank https://arche.readthedocs.io/en/jinja_report/nbs/In-Short.html (Coverage for scraped categories)

simoess commented:
> @simoess The last thing left - add a line to changes https://github.com/scrapinghub/arche/blob/jinja_report/CHANGES.md#037dev-work-in-progress. From the user perspective

Done

manycoding commented:
Thanks @simoess @ejulio @andersonberg , good job.
@ejulio I'll merge, it's not approved yet - your comments seem addressed to me.
Please tell me if otherwise
